proxy/proxy/client.go

```diff
@@ -444,6 +445,32 @@ func (c *Client) selectDialer() func(context.Context, string, string) (net.Conn,
 	}
 }
 
+func (c *Client) invalidateCfg(cfg *tls.Config, instance string) {
+	refreshCfgBuffer := c.RefreshCfgBuffer
+	if refreshCfgBuffer == 0 {
+		refreshCfgBuffer = DefaultRefreshCfgBuffer
+	}
+
+	c.cacheL.RLock()
+	e := c.cfgCache[instance]
+	c.cacheL.RUnlock()
+	if e.cfg != cfg || needsRefresh(e, refreshCfgBuffer) {
```

Comment:
After thinking about this a bit more, I think that my recommendation to use `needsRefresh` here is potentially problematic, because it doesn't account for when a new refresh has started "naturally" (based on time). For example:
1. The cert is ~5 mins from expiration, so a new refresh is triggered [here](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/60cde814b784f97482d0ff7f8221e725a08471be/proxy/proxy/client.go#L297)
1. the handshake errs, `invalidateCfg` is called, but `needsRefresh` returns `false` because a new refresh is already scheduled.

In this specific scenario, I _think_ we still want to invalidate `cfg` for the instance, as we want to block the next connect until the scheduled refresh is complete - not use the old (potentially invalid) info the short term. Does that sound right? 

If so, I think using `!isValid` would be better here than `needsRefresh`.

Reply:
If we remove the select statement as you suggested below, I think it should be sufficient to only check for `e.cfg != cfg` as this means that the config has either been invalidated or refreshed.

Reply:
I think at a high level, you only want to return early if:
1. The cfg has completed a refresh and something meaningful has changed (cert? ip?)
2. The cfg has already been invalidated (i.e., the next called to cachedCfg will block until the next refresh is complete)

The `e.cfg != cfg` seems potentially problematic in this scenario:
1. Goroutine A (_GA_): Encounters a handshake err, attempts to invalidate the config
2. GB: Calls `cachedCfg` when `needRefresh == true`, [updates `e.done and e.lastRefreshed`](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/60cde814b784f97482d0ff7f8221e725a08471be/proxy/proxy/client.go#L350)
3. GA: sees that `e.cfg != cfg` since `done` and `lastRefreshed` has changed, returns early
4. GA: Attempts to connect again, calls `cachedCfg`, `needsRefresh == false` since a new refresh was scheduled, but it also doesn't block on the new refresh (because `isValid == true`). Connect happens with the old certificate.

Reply:
> GA: sees that e.cfg != cfg since done and lastRefreshed has changed, returns early

I don't think this is the case as `e.cfg != cfg` compares the pointer address of the TLS config. This would return early only when a new `*tls.Config` is created (refresh completed) or when set to nil (invalidated).

Reply:
You're correct - I was thinking it was `cachedCfg`, not `*tls.Config`.